go-controller/pkg/ovn/egressfirewall.go

```diff
@@ -0,0 +1,328 @@
+		if splitStdout[index] == "" {
+			// between every entry there is a newline, skip the newline
+			index++
+		}
+		uuid := splitStdout[index]
+		var addressSetHash string
+		if strings.Contains(splitStdout[index+1], "ip4") {
+			addressSetHash = nsInfo.addressSet.GetIPv4HashName()
+		} else {
+			addressSetHash = nsInfo.addressSet.GetIPv6HashName()
+		}
+
+		for _, node := range existingNodes.Items {
+			_, stderr, err := util.RunOVNNbctl("remove", "logical_switch",
+				joinSwitch(node.Name), "acls", uuid)
+			if err != nil {
+				errList = append(errList, fmt.Errorf("failed to delete the rules for "+
+					"address_set=%s, stderr: %q (%v)", addressSetHash, stderr, err))
+			}
+
+		}
+	}
+	return errList
+}
+
+func (oc *Controller) updateEgressFirewallWithRetry(egressfirewall *egressfirewallapi.EgressFirewall) error {
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		return oc.kube.UpdateEgressFirewall(egressfirewall)
+	})
+}
+
+func (ef *egressFirewall) addACLToJoinSwitch(joinSwitch, hashedAddressSetNameIPv4, hashedAddressSetNameIPv6 string) error {
+	for _, rule := range ef.egressRules {
+		var match string
+		var action string
+		if rule.access == egressfirewallapi.EgressFirewallRuleAllow {
+			action = "allow"
+		} else {
+			action = "drop"
+		}
+
+		ipAddress, _, err := net.ParseCIDR(rule.to.cidrSelector)
+		if err != nil {
+			return fmt.Errorf("error rule.to.cidrSelector %s is not a valid CIDR (%+v)", rule.to.cidrSelector, err)
+		}
+		if !utilnet.IsIPv6(ipAddress) {
+			match = fmt.Sprintf("match=\"ip4.dst == %s && ip4.src == $%s\"", rule.to.cidrSelector, hashedAddressSetNameIPv4)
+		} else {
+			match = fmt.Sprintf("match=\"ip6.dst == %s && ip6.src == $%s\"", rule.to.cidrSelector, hashedAddressSetNameIPv6)
+		}
+
+		if len(rule.ports) > 0 {
+
+			match = fmt.Sprintf("%s && ( %s )\"", match[:len(match)-1], egressGetL4Match(rule.ports))
+		}
+		uuid, stderr, err := util.RunOVNNbctl("--data=bare", "--no-heading",
+			"--columns=_uuid", "find", "ACL", match, "action="+action,
+			fmt.Sprintf("external-ids:egressFirewall=%s", ef.namespace))
+
+		if err != nil {
+			return fmt.Errorf("error executing find ACL command, stderr: %q, %+v", stderr, err)
+		}
+		if uuid == "" {
+			_, stderr, err := util.RunOVNNbctl("--id=@acl", "create", "acl",
+				fmt.Sprintf("priority=%d", defaultStartPriority-rule.id),
+				fmt.Sprintf("direction=%s", fromLport), match, "action="+action,
+				fmt.Sprintf("external-ids:egressFirewall=%s", ef.namespace),
+				"--", "add", "logical_switch", joinSwitch,
+				"acls", "@acl")
+			if err != nil {
+				return fmt.Errorf("error executing create ACL command, stderr: %q, %+v", stderr, err)
+			}
+		} else {
+			_, stderr, err := util.RunOVNNbctl("add", "logical_switch", joinSwitch, "acls", uuid)
+			if err != nil {
+				return fmt.Errorf("error adding ACL to joinsSwitch %s failed, stderr: %q, %+v", joinSwitch, stderr, err)
+
+			}
+		}
+
```

Comment:
@JacobTanenbaum newline should get removed.

Reply:
done